Add examples for mic streaming (#26)
cobaltspeech/sdk-cubic: docs-src/content/getting-started/streaming.md

weight: 30
---
The following example shows how to transcribe an audio file using Cubic’s 
Streaming Recognize Request. The stream can come from a file on disk or be
directly from a microphone in real time.

<!--more-->

> ### Streaming from a local file

Reviewer: I prefer 'Streaming from an audio file'

> The examples below use a WAV file as input to the streaming recognition.

Reviewer: Here I think we should list again the supported file types - it's hidden in the intro but I think is worth repeating here

Reviewer 2: We should probably add the same list to the Synchronous Recognition page.

Author: I will add a list of types to the `streaming.md` file. If I recall correctly, we don't support the same list of formats for Synchronous Recognition (we expect raw samples currently). [See Issue](https://github.com/cobaltspeech/cubicsvr/issues/51)

```
{{% /tab %}}

{{%/tabs %}}

### Streaming from microphone

The examples require an external library to talk to the sound driver,
in order to stream audio from the microphone. Several libraries are
available but the recommended ones are mentioned for each language below.

{{%tabs %}}
{{% tab "Go" %}}

This example utilizes the [portaudio](https://github.com/gordonklaus/portaudio) bindings for Go
to stream audio from a microphone. To use this package, [PortAudio](http://www.portaudio.com/)
development headers and libraries to be installed. On Ubuntu this can be done using `apt-get install portaudio19-dev`.
Note that this package is specific to linux distributions and can't be used on Windows or Mac.

``` go
package main

import (
    "bytes"
    "context"
    "encoding/binary"
    "fmt"
    "log"

    "github.com/cobaltspeech/sdk-cubic/grpc/go-cubic"
    "github.com/cobaltspeech/sdk-cubic/grpc/go-cubic/cubicpb"
    "github.com/gordonklaus/portaudio"
)

const serverAddr = "127.0.0.1:2727"

// Microphone object which can be passed to Cubic's streaming
// recognition API as the audio stream. Two methods need to be

> // implemented for the Microphone object to : a Read method and

Reviewer 2: ```suggestion
// implemented for the Microphone object to be able to be passed to 
// StreamingRecognize: a Read method and
```

// a Close method. (See below)
type Microphone struct {
    buffer []int16
    stream *portaudio.Stream
}

// NewMicrophone instantiates a Microphone object with the desired
// sampling rate and buffer size. When streaming to cubic, the sample
// rate should be set to the sample rate of the model used. (See below)
func NewMicrophone(sampleRate, bufferSize uint32) (*Microphone, error) {

>     mic := Microphone{buffer: make([]int16, bufferSize/2)} // each sample = 2 bytes

Reviewer 2: Does `each sample = 2 bytes` depend on the sample rate?

Set a variable for `bufferSize/2` so you can use it here and on 167.

Author: No each sample just 16 bit = 2 byte. The cubic client measures it bufferSize in bytes while the PortAudio Microphone object deals in samples (be they 8bit or 16 bit etc.) So needed this conversion here to allow passing in of the same `bufferSize` used in `cubic.NewClient`

    portaudio.Initialize()
    stream, err := portaudio.OpenDefaultStream(1, 0, float64(sampleRate), int(bufferSize/2), mic.buffer)
    if err != nil {
        return nil, err
    }
    mic.stream = stream

    err = mic.stream.Start()
    if err != nil {
        return nil, err
    }

    return &mic, nil
}

// Read copies N bytes into the passed buffer from the microphone audio buffer

> // where N is the buffer size with which the cubic client was created.

Reviewer 2: ```suggestion
// where N is the buffer size passed to `cubic.NewClient`.
```

> // It also returns the number of bytes copied.

Reviewer 2: because if it reaches the end of the stream it might be less than `buffersize`?

Author: Yea, I think that maybe an issue when streaming from file, but for streaming from a microphone, we can always get a buffer full of bytes. It's just that the cubic client expects the `Read` method to return the number of bytes.

        else:
            print("\r{0}".format(result.alternatives[0].transcript), end="\n")

```
{{% /tab %}}

{{%/tabs %}}

### Streaming from microphone

The examples require an external library to talk to the sound driver,
in order to stream audio from the microphone. Several libraries are
available but the recommended ones are mentioned for each language below.

{{%tabs %}}
{{% tab "Go" %}}

This example utilizes the [portaudio](https://github.com/gordonklaus/portaudio) bindings for Go
to stream audio from a microphone. To use this package, install [PortAudio](http://www.portaudio.com/)
development headers and libraries using an appropriate package manager for your system (e.g. `apt-get install portaudio19-dev` on Ubuntu, `brew install portaudio` for OSX, etc.) or build from [source](http://portaudio.com/docs/v19-doxydocs/tutorial_start.html).

``` go
package main

import (
    "bytes"
    "context"
    "encoding/binary"
    "fmt"
    "log"

    "github.com/cobaltspeech/sdk-cubic/grpc/go-cubic"
    "github.com/cobaltspeech/sdk-cubic/grpc/go-cubic/cubicpb"
    "github.com/gordonklaus/portaudio"
)

const serverAddr = "127.0.0.1:2727"

> // Microphone object which can be passed to Cubic's streaming

Reviewer 3: Microphone implements the io.ReadCloser interface and provides a data stream for microphone input.

// recognition API as the audio stream. Two methods need to be
// implemented for the Microphone object to be able to be passed to
// StreamingRecognize: a Read method and
// a Close method. (See below)
type Microphone struct {
    buffer []int16
    stream *portaudio.Stream
}

// NewMicrophone instantiates a Microphone object with the desired
// sampling rate and buffer size. When streaming to cubic, the sample

> // rate should be set to the sample rate of the model used. (See below)

Reviewer 3: Remove "see below". It's not clear where to see :)

    mic := Microphone{buffer: make([]int16, numSamples)}

    portaudio.Initialize()
    stream, err := portaudio.OpenDefaultStream(1, 0, float64(sampleRate), int(numSamples), mic.buffer)
    if err != nil {
        return nil, err
    }
    mic.stream = stream

    err = mic.stream.Start()
    if err != nil {
        return nil, err
    }

    return &mic, nil
}

// Read copies N bytes into the passed buffer from the microphone audio buffer
// where N is the buffer size passed to `cubic.NewClient`. Also, to be compatible
// with the cubic client, it returns two things : an int representing the number of
// bytes copied and an error.
func (mic *Microphone) Read(buffer []byte) (int, error) {
    err := mic.stream.Read()
    if err != nil {
        return 0, err
    }
    byteBuffer := new(bytes.Buffer)
    err = binary.Write(byteBuffer, binary.LittleEndian, mic.buffer)
    if err != nil {
        return 0, err
    }
    copy(buffer, byteBuffer.Bytes())
    return len(buffer), nil
}

// Close shuts the microphone stream down and cleans up
func (mic *Microphone) Close() {
    mic.stream.Stop()
    mic.stream.Close()
    portaudio.Terminate()
}

func main() {

>     bufferSize := uint32(8192)  // 8192 bytes

Reviewer 3: remove the comment (redundant).

    client, err := cubic.NewClient(serverAddr, cubic.WithInsecure(), cubic.WithStreamingBufferSize(bufferSize))
    if err != nil {
        log.Fatal(err)
    }
    defer client.Close()

    modelResp, err := client.ListModels(context.Background())
    if err != nil {
        log.Fatal(err)
    }

    // Use the first available model
    model := modelResp.Models[0]

    // Create microphone stream
    mic, err := NewMicrophone(model.Attributes.SampleRate, bufferSize)
    if err != nil {
        log.Fatal(err)
    }
    defer mic.Close()

    cfg := &cubicpb.RecognitionConfig{
        ModelId: model.Id,
    }

    // Define a callback function to handle results
    resultHandler := func(resp *cubicpb.RecognitionResponse) {
        for _, r := range resp.Results {
            if r.IsPartial {
                fmt.Print("\r", r.Alternatives[0].Transcript) // print on same line
            } else {
                fmt.Println("\r", r.Alternatives[0].Transcript) // print and move to new line
            }
        }
    }

>     err = client.StreamingRecognize(context.Background(), cfg, mic, resultHandler)

Reviewer 3: move the microphone creation here, and add a comment that since mic implements io.ReadCloser we can use it directly in the StreamingRecognize function.

    if err != nil {
        log.Fatal(err)
    }
}
```
{{% /tab %}}

{{% tab "Python" %}}

This example requires the [pyaudio](http://people.csail.mit.edu/hubert/pyaudio/) 
module to stream audio from a microphone. Instructions for installing pyaudio for 
different systems are available at the link. On most platforms, this is simply `pip install pyaudio`.

``` python
import cubic
import pyaudio

serverAddress = '127.0.0.1:2727'

client = cubic.Client(serverAddress)

# get list of available models
modelResp = client.ListModels()

> for model in modelResp.models:

Reviewer 3: I think we can skip printing them here.

    print("ID = {}, Name = {}".format(model.id, model.name))

# use the first available model
model = modelResp.models[0]

cfg = cubic.RecognitionConfig(
    model_id = model.id
)

> # client.StreamingRecognize takes any binary

Reviewer 3: This comment seems to be wrapped a bit too short. Can you check and wrap it at col 80?
